db/change_cache_test.go

```diff
@@ -1433,3 +1433,161 @@ func TestLateArrivingSequenceTriggersOnChange(t *testing.T) {
+// sets query/cache boundaries
+func TestInitializeCacheUnderLoad(t *testing.T) {
+
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyChanges)()
+
+	// Increase the cache max size
+	cacheOptions := CacheOptions{}
+	cacheOptions.ChannelCacheMaxLength = 50
+
+	db, testBucket := setupTestDBWithCacheOptions(t, cacheOptions)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	db.ChannelMapper = channels.NewDefaultChannelMapper()
+
+	// Writes [docCount] documents.  Use wait group (writesDone)to identify when all docs have been written.
+	// Use another waitGroup (writesInProgress) to trigger getChanges midway through writes
+	docCount := 1000
+	inProgressCount := 100
+
+	var writesDone sync.WaitGroup
+	var writesInProgress sync.WaitGroup
+	writesDone.Add(docCount)
+	writesInProgress.Add(inProgressCount)
+
+	// Start writing docs
+	go func() {
+		for i := 0; i < docCount; i++ {
+			channels := []string{"zero"}
+			body := Body{"serialnumber": int64(i), "channels": channels}
+			docID := fmt.Sprintf("loadCache-%d", i)
+			_, err := db.Put(docID, body)
+			assert.NoError(t, err, "Couldn't create document")
+			if i < inProgressCount {
+				writesInProgress.Done()
+			}
+			writesDone.Done()
+		}
+	}()
+
+	// Wait for writes to be in progress, then getChanges for channel zero
+	writesInProgress.Wait()
+	changes, err := db.GetChanges(channels.SetOf(t, "zero"), ChangesOptions{})
+	assert.NoError(t, err, "Couldn't GetChanges")
+	firstChangesCount := len(changes)
+	var lastSeq SequenceID
+	if firstChangesCount > 0 {
+		lastSeq = changes[len(changes)-1].Seq
+	}
+
+	// Wait for writes to complete, then getChanges again
+	writesDone.Wait()
+	db.changeCache.waitForSequence(uint64(docCount), base.DefaultWaitForSequenceTesting, t)
+
+	changes, err = db.GetChanges(channels.SetOf(t, "zero"), ChangesOptions{Since: lastSeq})
+	assert.NoError(t, err, "Couldn't GetChanges")
+	secondChangesCount := len(changes)
+	assert.Equal(t, docCount, firstChangesCount+secondChangesCount)
+
+}
+
+// Verify that notifyChange for channel zero is sent even when the channel isn't active in the cache.
+func TestNotifyForInactiveChannel(t *testing.T) {
+
+	// Enable relevant logging
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyCache|base.KeyDCP)()
+
+	db, testBucket := setupTestDB(t)
+	defer tearDownTestDB(t, db)
+	defer testBucket.Close()
+
+	// -------- Setup notifyChange callback ----------------
+	changeCacheImpl := db.changeCache.(*changeCache)
+
+	var waitForNotify sync.WaitGroup
+	waitForNotify.Add(1)
+	changeCacheImpl.notifyChange = func(channels base.Set) {
+		log.Printf("channelsChanged: %v", channels)
+		if channels.Contains("zero") {
+			waitForNotify.Done()
+		}
+	}
+
+	// Write a document to channel zero
+	body := Body{"channels": []string{"zero"}}
+	_, err := db.Put("inactiveCacheNotify", body)
+	assert.NoError(t, err)
+
+	// Wait for notify
+	waitForNotify.Wait()
```

Comment:
Should we use the `WaitWithTimeout` util to prevent the test from hanging under a failure scenario?